Produce more accurate odometer fixes by averaging multiple GNSS poses while standing

rodja opened the pull request:
As an alternative to a Kalman Filter, this PR leverages the fact that we are standing still while the implement is working: 
We can collect multiple GNSS measurements and average them for a odometer pose update just before we start driving again. The implementation also ensures that the navigation pauses it's advancements shortly to gather a gps fix every 10 seconds for scenarios where the implement does not request stopping on their own.

pascalzauberzeug reviewed (commented):
Code looks good, looking forward to testing it

rodja commented:
Today, when testing with a real robot I discovered that we run into an endless loop if we do not have a RTK fix. This can happen indoors, when no internet connection is available, when SAPOS is not correctly configured or a thousand other causes. How should we handle this?

pascalzauberzeug commented:
@rodja do you mean the while-loop in update_robot_pose?

rodja commented:
Yes

pascalzauberzeug commented:
The `await self.gnss.update_robot_pose()` call should have a timeout and either we just continue to drive without an updated pose and warn the user about it or stop the automation.
I think the first option is better for our current use cases

rodja commented:
I agree. But before this PR, the default behaviour was simply to keep driving with odometry. Let's discuss this tomorrow in our all-hands meeting.

rodja commented:
As discussed we keep the old behaviour for now which does not require GNSS to be active. This allows us to drive by visual cues only.
By setting `system.gnss.ensure_gnss = True` this behaviour can later be changed by navigation strategies. The slippage-tests need it already. @pascalzauberzeug I think the code is ready for another real world test.

pascalzauberzeug commented:
The worked today on U4/RB28, but one time there was the following error where the automation was stopped and couldn't be started again without restarting the software.
```
rosys_1       | 2024-07-29 18:39:00.551 [INFO] rosys/rosys.py:69: automation stopped because an exception occurred in an automation
rosys_1       | 2024-07-29 18:39:00.552 [INFO] rosys/rosys.py:69: automation failed
rosys_1       | 2024-07-29 18:39:00.575 [ERROR] nicegui/app/app.py:119:
rosys_1       | Traceback (most recent call last):
rosys_1       |   File "/app/nicegui/background_tasks.py", line 52, in _handle_task_result
rosys_1       |     task.result()
rosys_1       |   File "/app/rosys/automation/automation.py", line 56, in __await__
rosys_1       |     signal = send(message)
rosys_1       |              ^^^^^^^^^^^^^
rosys_1       |   File "/app/field_friend/automations/navigation/navigation.py", line 50, in start
rosys_1       |     await self.gnss.update_robot_pose()
rosys_1       |   File "/app/field_friend/localization/gnss.py", line 119, in update_robot_pose
rosys_1       |     assert not self.is_paused
rosys_1       | AssertionError
```

rodja commented:
The tests with U4 worked very well after we fixed the `work_x` parameter. There also was an issue when pressing the stop button because of missing `super()` calls. This is now fixed. @pascalzauberzeug could you review the last changes? I think it's ready to merge!

pascalzauberzeug commented:
@rodja the field navigation tests fail after https://github.com/zauberzeug/field_friend/pull/130/commits/24d0da5f7701a71785a0f3022b2489e907b133cc.
this assert fails`assert system.field_navigation.automation_watcher.field_watch_active` in `test_approaching_first_row` and  `assert not system.automator.is_running` in `test_approaching_first_row_when_outside_of_field`.

I narrowed it down to the `await self.implement.activate()` in `FollowCropsNavigation.prepare()`

pascalzauberzeug commented:
https://github.com/zauberzeug/field_friend/blob/58a7354a486c17551a7b17a97d0b46e6c4626970/field_friend/automations/implements/recorder.py#L23
The test doesn't wait for this sleep. We could forward for x seconds or move `await super().activate()` to the end of the function and then forward until the implement is active
`await forward(until=lambda: system.current_implement.is_active)`

pascalzauberzeug commented:
Works again, will merge now!
